File: scripts/run_fixture_suite.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import os
import re
import time
from pathlib import Path

from compare_transcript import normalize
from faster_whisper import WhisperModel
from transcribe import DEFAULT_LANGUAGE, DEFAULT_MODEL, language_arg, resolve_compute_type, resolve_device


def slug(value: str) -> str:
    return re.sub(r"[^A-Za-z0-9_.-]+", "-", value).strip("-")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Run STT over a generated fixture suite and compare transcripts."
    )
    parser.add_argument("manifest", help="Fixture suite manifest JSON.")
    parser.add_argument(
        "--fixture-root",
        help="Generated fixture root. Default: fixtures/generated/<manifest id>",
    )
    parser.add_argument(
        "--model",
        default=os.environ.get("STT_MODEL", DEFAULT_MODEL),
        help=f"Whisper model name or local model path. Default: {DEFAULT_MODEL}",
    )
    parser.add_argument(
        "--language",
        default=os.environ.get("STT_LANGUAGE", DEFAULT_LANGUAGE),
        help="Language code such as ko, en, or auto. Default: ko",
    )
    parser.add_argument(
        "--device",
        choices=("auto", "cpu", "cuda"),
        default=os.environ.get("STT_DEVICE", "auto"),
        help="Inference device. Default: auto",
    )
    parser.add_argument(
        "--compute-type",
        default=os.environ.get("STT_COMPUTE_TYPE", "auto"),
        help="CTranslate2 compute type. Default: auto",
    )
    parser.add_argument(
        "--beam-size",
        type=int,
        default=int(os.environ.get("STT_BEAM_SIZE", "5")),
        help="Beam size for decoding. Default: 5",
    )
    parser.add_argument(
        "--model-dir",
        default=os.environ.get("STT_MODEL_DIR"),
        help="Optional model download/cache directory.",
    )
    parser.add_argument(
        "--output",
        help="Output JSON path. Default: output/suite/<manifest id>-<model>-<device>-<compute>.json",
    )
    parser.add_argument(
        "--require",
        choices=("exact", "normalized", "none"),
        default="normalized",
        help="Failure threshold. Default: normalized",
    )
    return parser.parse_args()


def load_manifest(path: str) -> dict:
    return json.loads(Path(path).read_text(encoding="utf-8"))


def fixture_dir(fixture_root: Path, row_idx: int) -> Path:
    return fixture_root / f"kss-row-{row_idx:05d}"


def transcribe_audio(model: WhisperModel, audio_file: Path, args: argparse.Namespace) -> tuple[str, float]:
    started_at = time.monotonic()
    segments, _ = model.transcribe(
        str(audio_file),
        language=language_arg(args.language),
        beam_size=args.beam_size,
        vad_filter=True,
        condition_on_previous_text=False,
    )
    parts = [segment.text.strip() for segment in segments if segment.text.strip()]
    return " ".join(parts).strip(), time.monotonic() - started_at


def result_ok(result: dict, required: str) -> bool:
    if required == "none":
        return True
    return bool(result[f"{required}_match"])


def main() -> int:
    args = parse_args()
    manifest = load_manifest(args.manifest)
    suite_id = manifest["id"]
    fixture_root = Path(args.fixture_root or f"fixtures/generated/{suite_id}")

    device = resolve_device(args.device)
    compute_type = resolve_compute_type(device, args.compute_type)
    output_path = Path(
        args.output
        or f"output/suite/{suite_id}-{slug(args.model)}-{device}-{compute_type}.json"
    )

    print(
        "loading model: "
        f"model={args.model} device={device} compute_type={compute_type}",
        flush=True,
    )
    model = WhisperModel(
        args.model,
        device=device,
        compute_type=compute_type,
        download_root=args.model_dir,
    )

    suite_started_at = time.monotonic()
    results = []
    for fixture in manifest["fixtures"]:
        row_idx = int(fixture["row_idx"])
        directory = fixture_dir(fixture_root, row_idx)
        audio_file = directory / "audio.wav"
        expected_file = directory / "expected.txt"
        if not audio_file.exists() or not expected_file.exists():
            raise FileNotFoundError(
                f"missing generated fixture for row {row_idx}: run scripts/fetch_kss_fixture.py --manifest {args.manifest}"
            )

        expected = expected_file.read_text(encoding="utf-8").strip()
        actual, elapsed = transcribe_audio(model, audio_file, args)
        exact_match = expected == actual
        normalized_match = normalize(expected) == normalize(actual)
        result = {
            "row_idx": row_idx,
            "label": fixture["label"],
            "expected": expected,
            "actual": actual,
            "exact_match": exact_match,
            "normalized_match": normalized_match,
            "elapsed": round(elapsed, 3),
        }
        results.append(result)
        status = "PASS" if result_ok(result, args.require) else "FAIL"
        print(
            f"{status} row={row_idx:05d} exact={exact_match} normalized={normalized_match} elapsed={elapsed:.2f}s"
        )

    summary = {
        "suite_id": suite_id,
        "manifest": args.manifest,
        "model": args.model,
        "device": device,
        "compute_type": compute_type,
        "required_match": args.require,
        "elapsed": round(time.monotonic() - suite_started_at, 3),
        "total": len(results),
        "exact_pass": sum(1 for result in results if result["exact_match"]),
        "normalized_pass": sum(1 for result in results if result["normalized_match"]),
        "results": results,
    }

    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(
        json.dumps(summary, ensure_ascii=False, indent=2) + "\n",
        encoding="utf-8",
    )
    print(f"output={output_path}")

    failures = [result for result in results if not result_ok(result, args.require)]
    return 1 if failures else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/app/cli/run-fixture-suite.ts
import { existsSync } from "node:fs";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { join } from "node:path";

import { compareTranscripts } from "@/features/transcript-comparison";
import { transcribeAudio } from "@/features/stt-adapter";
import { asRecord, readJsonFile } from "@/shared/json";

import { isDirectRun } from "./shared/direct-run";
import { parseArgs as parseTranscribeArgs } from "./transcribe";

interface ParsedArgs {
  manifest: string;
  fixtureRoot?: string;
  output?: string;
  require: "exact" | "normalized" | "none";
  transcribeArgs: string[];
}

export async function main(argv = process.argv.slice(2)): Promise<number> {
  const parsed = parseArgs(argv);
  if (parsed.kind === "help") {
    process.stdout.write(`${usage()}\n`);
    return 0;
  }
  if (parsed.kind === "error") {
    process.stderr.write(`error: ${parsed.message}\n${usage()}\n`);
    return 2;
  }

  try {
    const manifest = asRecord(
      await readJsonFile(parsed.args.manifest),
      "manifest must contain a JSON object",
    );
    const suiteId = requireString(manifest.id, "manifest id");
    const fixtures = requireArray(manifest.fixtures, "fixtures");
    const prefix =
      typeof manifest.fixture_dir_prefix === "string"
        ? manifest.fixture_dir_prefix
        : "kss-row";
    const fixtureRoot =
      parsed.args.fixtureRoot ?? join("fixtures/generated", suiteId);
    const output =
      parsed.args.output ??
      join("output/suite", `${suiteId}-typescript-adapter.json`);
    const results = [];
    const startedAt = performance.now();

    for (const rawFixture of fixtures) {
      const fixture = asRecord(rawFixture, "fixture must be an object");
      const rowIdx = requireNumber(fixture.row_idx, "row_idx");
      const directory = join(
        fixtureRoot,
        `${prefix}-${String(rowIdx).padStart(5, "0")}`,
      );
      const audioFile = join(directory, "audio.wav");
      const expectedFile = join(directory, "expected.txt");
      if (!existsSync(audioFile) || !existsSync(expectedFile)) {
        throw new Error(`missing generated fixture for row ${rowIdx}`);
      }

      const expected = (await readFile(expectedFile, "utf8")).trim();
      const transcribeParsed = parseTranscribeArgs([
        audioFile,
        ...parsed.args.transcribeArgs,
      ]);
      if (transcribeParsed.kind !== "ok") {
        throw new Error(
          transcribeParsed.kind === "error"
            ? transcribeParsed.message
            : "invalid transcribe args",
        );
      }
      const rowStartedAt = performance.now();
      const actual = await transcribeAudio(
        transcribeParsed.audioFile,
        transcribeParsed.options,
      );
      const exact = compareTranscripts({ expected, actual, exact: true }).ok;
      const normalized = compareTranscripts({ expected, actual }).ok;
      const result = {
        row_idx: rowIdx,
        label: fixture.label,
        category: fixture.category,
        cs_level: fixture.cs_level,
        expected,
        actual,
        exact_match: exact,
        normalized_match: normalized,
        elapsed: round3((performance.now() - rowStartedAt) / 1000),
      };
      results.push(result);
      const status = rowStatus(result, parsed.args.require);
      process.stdout.write(
        `${status} row=${String(rowIdx).padStart(5, "0")} exact=${exact} normalized=${normalized} elapsed=${result.elapsed.toFixed(2)}s\n`,
      );
    }

    const summary = {
      suite_id: suiteId,
      manifest: parsed.args.manifest,
      required_match: parsed.args.require,
      elapsed: round3((performance.now() - startedAt) / 1000),
      total: results.length,
      exact_pass: results.filter((result) => result.exact_match).length,
      normalized_pass: results.filter((result) => result.normalized_match)
        .length,
      results,
    };
    await mkdir(join(output, ".."), { recursive: true });
    await writeFile(output, `${JSON.stringify(summary, null, 2)}\n`, "utf8");
    process.stdout.write(`output=${output}\n`);
    return results.some((result) => !rowOk(result, parsed.args.require))
      ? 1
      : 0;
  } catch (error) {
    process.stderr.write(
      `error: ${error instanceof Error ? error.message : String(error)}\n`,
    );
    return 1;
  }
}

type ParseResult =
  | { kind: "ok"; args: ParsedArgs }
  | { kind: "help" }
  | { kind: "error"; message: string };

function parseArgs(argv: string[]): ParseResult {
  let manifest = "";
  let fixtureRoot: string | undefined;
  let output: string | undefined;
  let require: ParsedArgs["require"] = "normalized";
  const transcribeArgs: string[] = [];

  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index] ?? "";
    if (arg === "-h" || arg === "--help") return { kind: "help" };
    if (arg === "--fixture-root" || arg === "--output" || arg === "--require") {
      const value = argv[index + 1];
      if (!value) return { kind: "error", message: `${arg} requires a value` };
      index += 1;
      if (arg === "--fixture-root") fixtureRoot = value;
      else if (arg === "--output") output = value;
      else {
        if (value !== "exact" && value !== "normalized" && value !== "none") {
          return {
            kind: "error",
            message: "--require must be exact, normalized, or none",
          };
        }
        require = value;
      }
      continue;
    }
    if (arg.startsWith("-")) {
      transcribeArgs.push(arg, ...argv.slice(index + 1));
      break;
    }
    if (!manifest) {
      manifest = arg;
      continue;
    }
    transcribeArgs.push(arg, ...argv.slice(index + 1));
    break;
  }

  if (!manifest) return { kind: "error", message: "manifest is required" };
  return {
    kind: "ok",
    args: { manifest, fixtureRoot, output, require, transcribeArgs },
  };
}

function rowOk(
  result: { exact_match: boolean; normalized_match: boolean },
  required: ParsedArgs["require"],
): boolean {
  if (required === "none") return true;
  return required === "exact" ? result.exact_match : result.normalized_match;
}

function rowStatus(
  result: { exact_match: boolean; normalized_match: boolean },
  required: ParsedArgs["require"],
): string {
  if (required === "none") return "MEASURE";
  return rowOk(result, required) ? "PASS" : "FAIL";
}

function requireString(value: unknown, field: string): string {
  if (typeof value !== "string" || value.trim() === "") {
    throw new Error(`field '${field}' must be a non-empty string`);
  }
  return value;
}

function requireNumber(value: unknown, field: string): number {
  if (typeof value !== "number") {
    throw new Error(`field '${field}' must be a number`);
  }
  return value;
}

function requireArray(value: unknown, field: string): unknown[] {
  if (!Array.isArray(value)) {
    throw new Error(`field '${field}' must be a list`);
  }
  return value;
}

function round3(value: number): number {
  return Math.round(value * 1000) / 1000;
}

function usage(): string {
  return "Usage: npm run run-fixture-suite -- manifest.json [--model MODEL] [--device auto|cpu|cuda]";
}

if (isDirectRun(import.meta.url)) {
  process.exitCode = await main();
}
